src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -867,10 +909,19 @@ static void handle_rxflvl_ints(uint8_t rhport, USB_OTG_OUTEndpointTypeDef * out_
       xfer_ctl_t * xfer = XFER_CTL_BASE(epnum, TUSB_DIR_OUT);
 
       // Read packet off RxFIFO
-      read_fifo_packet(rhport, xfer->buffer, bcnt);
+      if (xfer->ff)
+      {
+        // Ring buffer
+        tu_fifo_write_n(xfer->ff, (const void *) rx_fifo, bcnt);
```

Comment:
That is true! It is possible to resolve this by implementing separate mutexes for write and read pointers. Technically, this would not be a problem but might be much work to do.

Reply:
@PanRe thanks to your awesome rework, I think the fifo can be written and read independently without conflict. If driver follow this rule firmly, maybe we could skip the mutex for it.

Reply:
I think so too yes, it anyway it would be weired to allow multiple processes to write into the fifo within one driver.

Reply:
@PanRe hmm, I forgot about multiple API() right across threads in application. Which is indeed true for cdc, the write()/print() can be called virtually across all the thread (with rtos). Hmmm, apparently we may need to put more thought on this, fifo does come with complexity of mutex though. We can separate write/read mutex for fifo and only enable mutex for the driver side (since mutex in isr can be tricky). I am not too worry about it right now, but definitely something to pay more attention later. Maybe this is not worth the effort for other non-timing sensitive driver, the middleman EP buff does a good job of isolation data.

Reply:
Well there is the case that you don't have a middle man buffer any more but only one FIFO... i think the best would be to extend the FIFO capabilities now and solve this issue once and for all by using separate mutexes for read and write and allow to not enforce use of mutexes just as you suggested!
Maybe the best would be to check within fifo.c if a mutex is set and only if so execute it. Otherwise just ignore it.
I can do this easily, the only nasty thing is that all projects where mutexes are used need to be updated...

Reply:
OK, mutex should be used whenever an preempted RTOS is used. Though you could do it in this PR or  the follow up one, whichever you prefer. Just make sure you put some `// TODO`  in the mutex access to remind us later on. I think only audio driver make use of the xfer_fifo for now.

Reply:
Wait, what exactly do you mean? Should
```
#if CFG_FIFO_MUTEX
static inline void tu_fifo_config_mutex(tu_fifo_t *f, tu_fifo_mutex_t mutex_hdl)
{
  f->mutex = mutex_hdl;
}
#endif
```
be changed to
```
#if CFG_FIFO_MUTEX
static inline void tu_fifo_config_mutex_write(tu_fifo_t *f, tu_fifo_mutex_t mutex_hdl)
{
  f->mutex_wr = mutex_hdl;
}
#endif
```
and
```
#if CFG_FIFO_MUTEX
static inline void tu_fifo_config_mutex_read(tu_fifo_t *f, tu_fifo_mutex_t mutex_hdl)
{
  f->mutex_rd = mutex_hdl;
}
#endif
```
? This would be a general solution but everywhere where `tu_fifo_config_mutex()` was used it needs to be changed...
Or do you mean i should not set mutexes in the audio driver at all?
There is another, however, rather nasty hack possible: we could introduce `tu_fifo_write_n_no_mutex()` etc. which ignores mutexes at all. This could be used in the dcd.c to read and write without being blocked. Then we do not break anything, however, it is rather ugly.

Reply:
@PanRe I mean you could do it now, or leave it to follow up PR. There is no need to implement it right away :). Regarding the API we can change it to 

```
tu_fifo_config_mutex(f, tu_fifo_mutex_t write_mutex, tu_fifo_mutex_t read_mutex)
```

and it is legit to set either as NULL or both Write and Read mutex to the same one, which we will do atm for existing driver . Regarding updating existing code, it should be no problem, tu_fifo is internal data type, there is only a handful of driver that use fifo and mutex. I wouldn't expect application to make use of tu_fifo.

> Or do you mean i should not set mutexes in the audio driver at all?

Sorry of not being clear, I only mean that could do mutex as follow up PR if wanted to.

> There is another, however, rather nasty hack possible: we could introduce tu_fifo_write_n_no_mutex() etc. which ignores mutexes at all. This could be used in the dcd.c to read and write without being blocked. Then we do not break anything, however, it is rather ugly.

I am against such as hack, we rather implement it later.

Reply:
Very well, i will implement it right away! It is just, we have discussed that stuff now, later we will have forgotten it ^^

Reply:
Superb! 👍 👍